### ISSUE TITLE
Compiling Arrow fails with "Unknown compiler" Apple LLVM version 9.0.0 on Mac OS

### ISSUE THREAD

**Issue by gycn** [ray-project/ray#1000]
The build seemed to fail when trying to compile Arrow with:

INFOCompiler id: AppleClang
CMake Error at cmake_modules/CompilerInfo.cmake:77 (message):
  Unknown compiler.  Version info:

  Apple LLVM version 9.0.0 (clang-900.0.37)

The full logs are attached.
[build_error.txt](https://github.com/ray-project/ray/files/1316493/build_error.txt)

**Comment by robertnishihara** [comment]
Interesting, maybe Arrow doesn't compile with clang 9. The relevant file is https://github.com/apache/arrow/blob/master/cpp/cmake_modules/CompilerInfo.cmake.

Want to try recompiling just arrow but adding an analogous block like 

```
# clang on Mac OS X, XCode 8.
elseif("${COMPILER_VERSION_FULL}" MATCHES ".*clang-8")
  set(COMPILER_FAMILY "clang")
  set(COMPILER_VERSION "3.8.0svn")
```

but for xcode 9 instead of xcode 8?

To recompile you'll probably need to do something like

```
cd arrow/cpp/build
rm -rf *
cmake ..
make -j8
```

No idea if that will work, but it's worth a try.

**Comment by gycn** [comment]
I added this block and it worked.

# clang on Mac OS X, XCode 9. Thank you
```
elseif("${COMPILER_VERSION_FULL}" MATCHES ".*clang-9")
   set(COMPILER_FAMILY "clang")
    set(COMPILER_VERSION "3.8.0svn")
```

**Comment by gycn** [comment]
Now it's throwing another error (logs attached). I'll try to see if I can solve this.
[build_error2.txt](https://github.com/ray-project/ray/files/1319469/build_error2.txt)

**Comment by xhochy** [comment]
Opened https://issues.apache.org/jira/browse/ARROW-1591 about that. I can reproduce locally and should be able to fix this quite quickly.

**Comment by robertnishihara** [comment]
Thanks @xhochy!

**Comment by robertnishihara** [comment]
@gycn, can you try out the fix in apache/arrow#1121 and let me know if that works for you?

You can just try to build the current arrow master.

**Comment by gycn** [comment]
I was able to get Arrow to build, but now there's a problem with Plasma. Will update once I verify what the issue is.

**Comment by robertnishihara** [comment]
Want to try the workaround in this issue? https://github.com/tensorflow/tensorflow/issues/13220#issuecomment-331579775

**Comment by gycn** [comment]
I was finally able to get it to build by including the snippet below in src/thirdparty/arrow/cpp.src/plasma.h, src/thirdparty/arrow/cpp.src/common.h, and src/thirdparty/arrow/cpp.src/io.h

#ifndef _MACH_PORT_T
#define _MACH_PORT_T
#include <sys/_types.h> /* __darwin_mach_port_t */
typedef __darwin_mach_port_t mach_port_t;
#include <pthread.h>
mach_port_t pthread_mach_thread_np(pthread_t);
#endif /* _MACH_PORT_T */

**Comment by robertnishihara** [comment]
Nice! Want to submit a PR to Arrow? Does Ray work for you after making that change?

**Comment by gycn** [comment]
Yep, Ray seems to work fine. I made a pull request here: https://github.com/apache/arrow/pull/1139.

**Comment by robertnishihara** [comment]
Fixed by #1042.
